harness/determined/pytorch/dsat/_dsat_search_method.py

```diff
@@ -845,55 +891,49 @@ def get_trial_list_after_successful_run(
-        should_stop_this_stage_3_trial = last_trial.stage == 3 and stage_one_or_two_successful
-        if should_stop_this_stage_3_trial:
-            return True
+        should_stop_this_stage_3_trial = trial.stage == 3 and stage_one_or_two_successful
 
-        if last_trial.error:
-            failed_on_minimal_batch = last_trial.mbs == last_trial.search_data["lo"]
-            if failed_on_minimal_batch:
-                return True
+        # Check if other same-stage trials have successfully run with larger batch sizes than this
+        # lineage can possibly run.
 
-            successful_configs_at_same_mbs_and_stage = any(
-                trial.mbs >= last_trial.mbs
-                for request_id, trial in self.trial_tracker
-                if request_id != last_trial.request_id
-                and trial.stage == last_trial.stage
-                and trial.searcher_metric_val
-            )
-            if successful_configs_at_same_mbs_and_stage:
-                return True
-
-        elif not last_trial.error:
-
-            def better(trial1, trial2):
-                if self.trial_tracker.smaller_is_better:
-                    return trial1.searcher_metric_val < trial2.searcher_metric_val
-                else:
-                    return trial1.searcher_metric_val > trial2.searcher_metric_val
+        other_configs_run_larger_batch_sizes = trial.error_in_direct_history and any(
+            other_trial.mbs >= trial.search_data.hi
+            for _, other_trial in self.trial_tracker
+            if other_trial.stage == trial.stage and other_trial.searcher_metric_val is not None
+        )
 
-            better_configs_at_same_mbs_and_stage = any(
-                better(trial, last_trial)
-                for request_id, trial in self.trial_tracker
-                if request_id != last_trial.request_id
-                and trial.stage == last_trial.stage
-                and trial.searcher_metric_val
-            )
-            if better_configs_at_same_mbs_and_stage:
-                return True
+        if (
+            failed_on_min_mbs
+            or ran_largest_possible_mbs
+            or exceeded_trials_per_random_config_limit
+            or should_stop_this_stage_3_trial
+            or other_configs_run_larger_batch_sizes
+        ):
+            return True
 
         return False
 
     def get_random_mbs_from_search_data(self, search_data: Dict[str, int]) -> int:
-        mbs = search_data["lo"] + self.rng.binomial(search_data["hi"] - search_data["lo"], 0.5)
-        assert search_data["lo"] <= mbs <= search_data["hi"]  # TODO: remove
+        """
+        Randomly choose a mbs given the `search_data` bounds. Random choice covers a larger search
+        volume than simply choosing the midpoint. Draws from a binomial distribution, to keep the
+        results still somewhat focused near the midpoint.
+        """
+        mbs = search_data.lo + self.rng.binomial(search_data.hi - search_data.lo, 0.5)
```

Comment:
Just checking my understanding, why must we randomize the midpoint check? Is it because we might have multiple lineages with the same mbs range but same stage and don't want to repeat work? If that is the case, why make these duplicate lineages in the first place?

Reply:
It's a small optimization, based on the general fact that you do a better job exploring parameter space with a random search than with a grid search, and using the strict midpoint would be more like a grid search.

Phenomenologically, this also just appeared to perform better in testing, though it wasn't very systematic.